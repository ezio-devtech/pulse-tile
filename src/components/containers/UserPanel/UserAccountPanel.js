import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { lifecycle } from 'recompose';
import { get } from 'lodash';

import packageJSON from '../../../../package.json';
import { testConstants, isDevMode } from '../../../config/for-test.constants';
import PTButton from '../../ui-elements/PTButton/PTButton';
import UserAccountPanelSettings from './UserAccountPanelSettings';
import userAccountSelector from './selectors';
import { initialiseSelector } from '../App/selectors';
import { clientUrls } from '../../../config/client-urls.constants';
import { unmountOnBlur } from '../../../utils/HOCs/unmount-on-blur.utils';
import { logoutStart } from '../../../ducks/logout.duck';
import { themeConfigs } from '../../../themes.config';
import { getDDMMMYYYY } from '../../../utils/time-helpers.utils';

const mapDispatchToProps = dispatch => ({ actions: bindActionCreators({ logoutStart }, dispatch) });

@connect(initialiseSelector)
@connect(userAccountSelector, mapDispatchToProps)
@lifecycle(unmountOnBlur)
export default class UserAccountPanel extends PureComponent {

  static propTypes = {
    user: PropTypes.shape().isRequired,
    actions: PropTypes.objectOf(PropTypes.func).isRequired,
  };

  static contextTypes = {
    router: PropTypes.shape({
      history: PropTypes.object,
    }),
  };

  renderToProfile() {
    this.context.router.history.push(clientUrls.USER_PROFILE);
  }

  render() {
    const { user, actions, onClick, onClose, initialiseData } = this.props;
    const varsionOfPulseTile = packageJSON.version;
    const varsionOfReact = packageJSON.dependencies.react;
    const imageLocation = '/images/user.jpg';
    const imageSource = isDevMode ? (testConstants.hostName + imageLocation) : imageLocation;
    return (
      <div className="dropdown-user dropdown-menu-right dropdown-menu" id="userAccountPanelElement" >
        { themeConfigs.isShowUserPhoto ?
            <div className="user-profile-image" onClick={() => this.renderToProfile()}>
              <div className="img">
                <img src={imageSource} alt="" />
              </div>
            </div>
        : null }
        <div className="user-profile-info">
          <div
            className="user-profile-info__item name"
            onClick={() =>{
              this.context.router.history.push(clientUrls.USER_PROFILE);
              onClick('')
            }}
          >{user.given_name} {user.family_name}</div>
          <div className="user-profile-info__descr">
            <div className="user-profile-info__item role">User Role:{user.role}</div>
            <div className="user-profile-info__item email">{user.email}</div>
              {(get(themeConfigs, 'isShowUserBirthday', false) && get(user, 'dateOfBirth', null))
              ? <div className="user-profile-info__item birthday">Date of Birth: {getDDMMMYYYY(user.dateOfBirth)}</div>
              : null
            }
          </div>
          { themeConfigs.isShowUserProfileSpecification ?
            <div className="specification">
              <div className="user-profile-info__title registered">Registered GP</div>
              <div className="user-profile-info__item version-back">About Showcase Stack; PulseTile version {varsionOfPulseTile}/QEWD_Ripple version {initialiseData.version}</div>
              <div className="user-profile-info__item version-front">React version {varsionOfReact}</div>
            </div>
            : null }
          <PTButton className="btn btn-theme btn-block btn-signout" aria-label="Log Out" onClick={actions.logoutStart} accessKey="S">
            <div>
              <span className="brn-text">Sign Out</span> <i className="btn-icon fa fa-sign-out" />
            </div>
          </PTButton>
        </div>
        { themeConfigs.isShowUserProfileSettings && <UserAccountPanelSettings onClose={onClose} /> }
      </div>
    )
  }
}
