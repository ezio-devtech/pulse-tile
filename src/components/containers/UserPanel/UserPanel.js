import React, { PureComponent } from 'react';
import classNames from 'classnames';
import { get } from 'lodash';

import PTButton from '../../ui-elements/PTButton/PTButton';
import UserPanelItem from './UserPanelItem';
import NotificationContent from '../../presentational/temprorary/NotificationContent'
import UserAccountPanel from './UserAccountPanel'
import { themeConfigs } from '../../../themes.config';

import TopHeaderButtons from '../../theme/components/TopHeaderButtons';

const USER_ACCOUNT_PANEL = 'userAccountPanel';
const NOTIFICATION_CONTENT = 'notificationContent';

export default class UserPanel extends PureComponent {

  static defaultProps = {
    addUserPanels: [],
    pageUrl: '',
    userId: '',
  };

  state = {
    openedPanel: '',
  };

  /* istanbul ignore next */
  componentWillMount() {
    document.addEventListener('click', this.handleClick, false);
  }

  /* istanbul ignore next */
  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick, false);
  }



  handleClick = /* istanbul ignore next */ (e) => {
    if (!this.node.contains(e.target)) {
      this.setState({ openedPanel: '' });
    }
  };

  handleMouseDown = (name) => {
    /* istanbul ignore next */
    this.setState((prevState) => {
      if (prevState.openedPanel !== name) {
        return ({ openedPanel: name })
      }
      return ({ openedPanel: '' })
    })
  };

  closePanel = () => {
    this.setState({ openedPanel: '' });
  };

  getUserPanelsItems = (addUserPanels) => {
    if (addUserPanels.length) {
      return addUserPanels.map((el) => {
        return (
          <UserPanelItem className="user-panel-item">
            { el }
          </UserPanelItem>
        );
      });
    }
    return null;
  };

  render() {
    const { openedPanel } = this.state;
    const { addUserPanels, pageUrl, homepageLink } = this.props;


    const isSearch = get(themeConfigs, 'topHeader.showSearch', true);
    const isNotifications = get(themeConfigs, 'topHeader.showNotifications', true);
    const isUserPanel = get(themeConfigs, 'topHeader.showUserPanel', true);

    const additionalUserPanels = this.getUserPanelsItems(addUserPanels);

    return (
      <ul className="user-panel" role="tablist" ref={node => this.node = node}>
        { additionalUserPanels }

        <TopHeaderButtons pageUrl={pageUrl} homepageLink={homepageLink} />

        {isNotifications ? <UserPanelItem role="tab" className={classNames('user-panel-item dropdown', { 'open': openedPanel === NOTIFICATION_CONTENT })}>
          <NotificationContent />
          <PTButton className="btn-header btn-notification" onClick={() => this.handleMouseDown(NOTIFICATION_CONTENT)}>
            <div>
              <i className="fa fa-bell-o" />
              <span className="count">2</span>
            </div>
          </PTButton>
        </UserPanelItem> : null}

        {isUserPanel ? <UserPanelItem role="tab" className={classNames('user-panel-item dropdown', { 'open': openedPanel === USER_ACCOUNT_PANEL })}>
          <UserAccountPanel onClick={this.handleMouseDown} onClose={this.closePanel} />
          <PTButton aria-controls="userAccountPanelElement" id="icon-profile" className="btn-header btn-user" onClick={() => this.handleMouseDown(USER_ACCOUNT_PANEL)}>
            <i className="fa fa-user" />
          </PTButton>
        </UserPanelItem> : null}

      </ul>
    )
  }
}
