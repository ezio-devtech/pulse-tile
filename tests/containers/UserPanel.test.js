import React from 'react';
import Enzyme, { mount, render, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import { Provider } from 'react-redux'
import { StaticRouter } from 'react-router'
import configureStore from 'redux-mock-store';

import UserPanel from '../../src/components/containers/UserPanel/UserPanel';

Enzyme.configure({ adapter: new Adapter() })

const mockStore = configureStore();
const store = mockStore({
  runtime: {},
  userAccount: {
    sub: '28AD8576-1948-4C84-8B5E-55FB7EE027CE',
    given_name: 'Bob',
    family_name: 'Smith',
    email: 'bob.smith@gmail.com',
    tenant: null,
    role: 'IDCR',
    roles: [
      'IDCR',
    ],
  },
});
const context = {};

describe('Component <UserPanel />', () => {
  it('should renders with props correctly', () => {
    const userPanel = mount(
      <Provider store={store}>
        <StaticRouter location="someLocation" context={context}>
          <UserPanel
            openedPanel="userAccountPanel"
          />
        </StaticRouter>
      </Provider>)
    expect(userPanel).toMatchSnapshot();
  });
  it('should renders with props correctly', () => {
    const userPanel = shallow(
      <UserPanel
        store={store}
      />);
    expect(userPanel).toMatchSnapshot();
    expect(userPanel.state().openedPanel).toEqual('');

    userPanel.find('.btn-notification').simulate('click');
    expect(userPanel.state().openedPanel).toEqual('notificationContent');

    userPanel.setState({ openedPanel: 'userAccountPanel' });
    expect(userPanel.state().openedPanel).toEqual('userAccountPanel');

    userPanel.find('.btn-notification').simulate('click');
    userPanel.find('.btn-user').simulate('click');
    expect(userPanel.state().openedPanel).toEqual('userAccountPanel');
  });
});
