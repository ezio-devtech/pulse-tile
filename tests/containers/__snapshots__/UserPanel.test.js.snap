// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <UserPanel /> should renders with props correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <StaticRouter
    basename=""
    context={Object {}}
    location="someLocation"
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "someLocation",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <UserPanel
        addUserPanels={Array []}
        isNotifications={true}
        isQuestions={true}
        isSearch={true}
        isUserPanel={true}
        openedPanel="userAccountPanel"
      >
        <ul
          className="user-panel"
          role="tablist"
        >
          <UserPanelItem
            className="user-panel-item visible-xs"
          >
            <li
              className="user-panel-item visible-xs"
            >
              <PTButton
                className="btn-header"
              >
                <button
                  className="btn-header"
                >
                  <i
                    className="fa fa-search"
                  />
                </button>
              </PTButton>
            </li>
          </UserPanelItem>
          <UserPanelItem
            className="user-panel-item"
          >
            <li
              className="user-panel-item"
            >
              <PTButton
                className="btn-header"
              >
                <button
                  className="btn-header"
                >
                  <i
                    className="fa fa-question-circle"
                  />
                </button>
              </PTButton>
            </li>
          </UserPanelItem>
          <UserPanelItem
            className="user-panel-item dropdown"
          >
            <li
              className="user-panel-item dropdown"
            >
              <NotificationContent>
                <div
                  className="notifications-drop dropdown-menu dropdown-menu-right"
                  id="tab-notifications"
                >
                  <div
                    className="notifications-header"
                  >
                    Notifications (2)
                  </div>
                  <div
                    className="notifications-body"
                  >
                    <ul
                      className="notifications-list"
                    >
                      <li
                        className="notifications-list-item"
                      >
                        <div
                          className="notification-info"
                        >
                          Information from Liquid Logic (Social Care) is currently unavailable.
                        </div>
                        <div
                          className="notification-time"
                        >
                          <em>
                            10 minutes ago
                          </em>
                        </div>
                      </li>
                      <li
                        className="notifications-list-item"
                      >
                        <div
                          className="notification-info"
                        >
                          Information from Liquid Logic (Social Care) is currently unavailable.
                        </div>
                        <div
                          className="notification-time"
                        >
                          <em>
                            15 minutes ago
                          </em>
                        </div>
                      </li>
                    </ul>
                  </div>
                </div>
              </NotificationContent>
              <PTButton
                className="btn-header btn-notification"
                onClick={[Function]}
              >
                <button
                  className="btn-header btn-notification"
                  onClick={[Function]}
                >
                  <div>
                    <i
                      className="fa fa-bell-o"
                    />
                    <span
                      className="count"
                    >
                      2
                    </span>
                  </div>
                </button>
              </PTButton>
            </li>
          </UserPanelItem>
          <UserPanelItem
            className="user-panel-item dropdown"
          >
            <li
              className="user-panel-item dropdown"
            >
              <Connect(lifecycle(UserAccountPanel))
                onClick={[Function]}
              >
                <lifecycle(UserAccountPanel)
                  actions={
                    Object {
                      "logoutStart": [Function],
                    }
                  }
                  onClick={[Function]}
                  user={
                    Object {
                      "email": "bob.smith@gmail.com",
                      "family_name": "Smith",
                      "given_name": "Bob",
                      "role": "IDCR",
                      "roles": Array [
                        "IDCR",
                      ],
                      "sub": "28AD8576-1948-4C84-8B5E-55FB7EE027CE",
                      "tenant": null,
                    }
                  }
                >
                  <UserAccountPanel
                    actions={
                      Object {
                        "logoutStart": [Function],
                      }
                    }
                    onClick={[Function]}
                    user={
                      Object {
                        "email": "bob.smith@gmail.com",
                        "family_name": "Smith",
                        "given_name": "Bob",
                        "role": "IDCR",
                        "roles": Array [
                          "IDCR",
                        ],
                        "sub": "28AD8576-1948-4C84-8B5E-55FB7EE027CE",
                        "tenant": null,
                      }
                    }
                  >
                    <div
                      className="dropdown-user dropdown-menu-right dropdown-menu"
                    >
                      <div
                        className="user-profile-image"
                        onClick={[Function]}
                      >
                        <div
                          className="img"
                        >
                          <img
                            alt=""
                            src="user.jpg"
                          />
                        </div>
                      </div>
                      <div
                        className="user-profile-info"
                      >
                        <div
                          className="name"
                          onClick={[Function]}
                        >
                          Bob
                           
                          Smith
                        </div>
                        <div
                          className="specification"
                        >
                          <div
                            className="item"
                          >
                            <em>
                              IDCR
                            </em>
                          </div>
                          <div
                            className="item"
                          >
                            bob.smith@gmail.com
                          </div>
                          <div
                            className="item gray"
                          >
                            <em>
                              10/05/2099
                            </em>
                          </div>
                          <div
                            className="item"
                          >
                            <em>
                              About Showcase Stack; PulseTile version 1.0.0/QEWD_Ripple version 1.0.0
                            </em>
                          </div>
                        </div>
                        <PTButton
                          className="btn btn-success btn-block btn-signout"
                        >
                          <button
                            className="btn btn-success btn-block btn-signout"
                          >
                            <div
                              onClick={[Function]}
                            >
                              <span
                                className="brn-text"
                              >
                                Sign Out
                              </span>
                               
                              <i
                                className="btn-icon fa fa-sign-out"
                              />
                            </div>
                          </button>
                        </PTButton>
                      </div>
                    </div>
                  </UserAccountPanel>
                </lifecycle(UserAccountPanel)>
              </Connect(lifecycle(UserAccountPanel))>
              <PTButton
                className="btn-header btn-user"
                onClick={[Function]}
              >
                <button
                  className="btn-header btn-user"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-user"
                  />
                </button>
              </PTButton>
            </li>
          </UserPanelItem>
        </ul>
      </UserPanel>
    </Router>
  </StaticRouter>
</Provider>
`;

exports[`Component <UserPanel /> should renders with props correctly 2`] = `
<ul
  className="user-panel"
  role="tablist"
>
  <UserPanelItem
    className="user-panel-item visible-xs"
  >
    <PTButton
      className="btn-header"
    >
      <i
        className="fa fa-search"
      />
    </PTButton>
  </UserPanelItem>
  <UserPanelItem
    className="user-panel-item"
  >
    <PTButton
      className="btn-header"
    >
      <i
        className="fa fa-question-circle"
      />
    </PTButton>
  </UserPanelItem>
  <UserPanelItem
    className="user-panel-item dropdown"
  >
    <NotificationContent />
    <PTButton
      className="btn-header btn-notification"
      onClick={[Function]}
    >
      <div>
        <i
          className="fa fa-bell-o"
        />
        <span
          className="count"
        >
          2
        </span>
      </div>
    </PTButton>
  </UserPanelItem>
  <UserPanelItem
    className="user-panel-item dropdown"
  >
    <Connect(lifecycle(UserAccountPanel))
      onClick={[Function]}
    />
    <PTButton
      className="btn-header btn-user"
      onClick={[Function]}
    >
      <i
        className="fa fa-user"
      />
    </PTButton>
  </UserPanelItem>
</ul>
`;
